### PR TITLE
Emit event when frame detaches

### PR DESCRIPTION
**Pull request by imurchie**
Some minor changes. The substantive change is to emit an event when frames detach. This will allow us to void the frames in the drivers.

### CODE REVIEW

**Comment by imurchie** [comment]
I am rethinking frame handling in general, but this will fix things in the interim.

File: lib/remote-debugger.js [hunk 1]
```diff
@@ -245,9 +246,16 @@ class RemoteDebugger extends events.EventEmitter {
...
           let found = false;
-          dictLoop: for (let appDict of _.values(this.appDict)) {
+          dictLoop: for (const appDict of _.values(this.appDict)) {
             if (found) break; // eslint-disable-line curly
-            for (let dict of (appDict.pageDict || [])) {
+
+            // if the page dictionary has not been loaded yet from the web
+            // inspector, wait for it or time out after 10s
+            if (appDict.pageDict.promise) {
+              await Promise.any([Promise.delay(10000), appDict.pageDict.promise]);
```
Comment (mykola-mokhnach) [inline lib/remote-debugger.js]: shall we ignore the result if delay wins the race?

Comment (mykola-mokhnach) [reply lib/remote-debugger.js]: we could also do something like `await Promise.resolve(appDict.pageDict.promise).timeout(...)`

File: lib/remote-debugger.js [hunk 2]
```diff
@@ -245,9 +246,24 @@ class RemoteDebugger extends events.EventEmitter {
...
             if (found) break; // eslint-disable-line curly
-            for (let dict of (appDict.pageDict || [])) {
+
+            // if the page dictionary has not been loaded yet from the web
+            // inspector, wait for it or time out after 10s
+            if (appDict.pageDict.promise) {
+              try {
+                await Promise.resolve(appDict.pageDict.promise).timeout(10000);
+              } catch (err) {
+                if (!(err instanceof Promise.TimeoutError)) {
```
Comment (mykola-mokhnach) [inline lib/remote-debugger.js]: does instanceof work as expected here or it would be better to use an explicit `constructor.name` comparison?

Comment (imurchie) [reply lib/remote-debugger.js]: It does work.

**Pull request merged**